homeassistant/components/firmata/pin.py

```diff
@@ -0,0 +1,145 @@
+"""Code to handle pins on a Firmata board."""
+import logging
+
+from homeassistant.core import callback
+
+from .const import PIN_MODE_INPUT, PIN_MODE_PULLUP
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class FirmataPinUsedException(Exception):
+    """Represents an exception when a pin is already in use."""
+
+
+class FirmataBoardPin:
+    """Manages a single Firmata board pin."""
+
+    def __init__(self, board, pin, pin_mode):
+        """Initialize the pin."""
+        self._board = board
+        self._pin = pin
+        self._pin_mode = pin_mode
+        self._pin_type, self._firmata_pin = self._board.get_pin_type(self._pin)
+        self._state = None
+
+        if not self._board.mark_pin_used(self._pin):
```

Comment:
We may want to move this to a dedicated setup method since we don't want side effects in init method.

Reply:
If the pin class initialization is moved outside of the entity then it can be called there (requested change above) in a separate function. I guess that would resolve this since then the entity would never be created if there is a pin conflict.

Reply:
We still don't want side effects in the init method here. There's technically a side effect in the `mark_pin_used` method.